Bug fix: generate FieldnameWithTypeAndReference for dictionary fields.

romkatv opened the pull request.

romkatv commented:
I wasn't able to run tests. Killed half an hour trying every suggestion from StackOverflow but Visual Studio still didn't see any tests.

kbilsted commented:
Thanks for the pull request. Currently there are failing tests. Are you using VS with resharper 10? I have major issues myself with resharper 10 and detecting tests. Just downgraded to resharper 9 for that very reason.

romkatv commented:
Hi Kasper,

Thanks for the prompt reply. In retrospect, pull request wasn't the right medium. I should've filed a feature request -- I believe it would work better. Let me try to explain what I'm doing.

I've been happily using my own formatter that produces strings like these:

  {A = 0, B = {C = 1, D = 2}}

Lists look like this:

  {A = ([0] = "hello", [1] = "bye")}

I'd like dictionaries to look similar to lists:

  {A = ([42] = "hello", [1337] = "bye")}

I managed to get it working but it's really awkward because there is no FieldnameWithTypeAndReference token when the dictionary starts.

Does that make sense? Do you think the library could support this type of formatting?

Another and potentially bigger problem is that it's not possible to print empty dictionaries. Consider the following class definition:

class X {
  public Dictionary<int, int> D = new Dictionary<int, int>();
};

Introspection generates the following tokens for an instance of X with empty D:
1. FieldnameWithTypeAndReference for "X".
2. StartScope.
3. StartEnumeration.
4. EndEnumeration.
5. EndScope.

The dictionary is described with StartEnumeration and EndEnumeration, but neither of them has the the field name -- "D".

kbilsted commented:
Hi @romkatv thank you for the clarification. I think it sounds reasonable. Also the json output is not really json so that need to be reworked anyway.so yeah we can just as well introduce breaking changes :)

kbilsted commented:
If you still cannot run the tests feel free to upgrade the nunit framework.  I've noticed that in the v3 there no longer is a message that you can supply with the assert or maybe I'm overlooking something.  That may prove a challenge for the api.. or maybe a wrapper function can be exposed to wrap and ignore passing in the message parameter.

romkatv commented:
I already tried upgrading to NUnit 3. Didn't help.

Perhaps it would be easier for you to update the test cases, provided that my fix makes sense and you have a way to automatically update the test cases? If not, I'll try to get tests working on my machine when I get a free minute (probably next weekend).

kbilsted commented:
Try installing the ugly nunit gui runner . If I upgrade nunit I have to upgrade resharper also. And v10 suxx..

We could change it to xunit but I dislike how you have to run through hoops in order to do casual console write

romkatv commented:
OK, I got it running with the nunit gui runner. Thanks for the suggestion!

The JSON and XML formatters need fixing. I'll ping you when I'm done with them. Cheers!

kbilsted commented:
Great. But it's fine with me to take small incremental changes. Eg a pr per format.

I'm also considering if making all the processes more lazy could speed up things.  But my experience is that it sacrifices debugging

romkatv commented:
I've made a few more changes. Please take another look.

1. Split Enumeration into two distinct sequence types: List and Dict. Instead of TokenType.{Start,Stop}Enumeration we now have TokenType.{Start,Stop}{List,Dict}.
2. Split Field.SimpleKeyInArrayOrDictionary into Field.Key (for elements of a Dict) and Field.Index (for elements of a List). Coupled with (1) it makes it easier to write formatters that distinguish between lists and dicts (which is pretty much all of them).
3. Made JSON formatter more similar to JSON.
4. Made XML formatter produce less verbose output with more consistent schema.
5. Updated all tests.

kbilsted commented:
hi @romkatv  thanks. I'll have a look at it.Without having looked at the code I feel that the list-type you mentioned perhaps is better named as "enumerable" or something - it may be far from a list, but it is something whose elements are not organized as key,value

you also say that it is closer to JSON, what is still outstanding?

romkatv commented:
> you also say that it is closer to JSON, what is still outstanding?

There are a few cases where the produced output isn't valid JSON. For example, string fields may have quotes in them which don't get escaped (XML formatter has the same problem but with different symbols). Then references aren't part of the JSON spec, of course. I mean, it's good enough for ToString(), which is only meant to produce human-readable output.

There is also weird double-quoting of strings in XML format. That's not how XML usually looks like. Actually, let me just fix that.... Done.

romkatv commented:
> I feel that the list-type you mentioned perhaps is better named as "enumerable" or something

Right, I used 'list' in a sense it's often used in other languages, meaning a sequence. In C# it corresponds to IEnumerable. I thought that calling it StartEnumerable and EndEnumerable might be confusing because not all IEnumerable instances generate these (string and Dictionary with simple keys don't). By the way, I found StartEnumeration confusing, too -- initially I thought it's about the enums.

Anyway, that's why I chose these names but I'm not attached to them. You are the boss, let me know what you want them to be called.

kbilsted commented:
@romkatv list vs enumerable:  well I think I agree with you - I just felt we needed that discussion brought to the table.

kbilsted commented:
Was a busy day at work today - I hope to have more energy tomorrow I'll scrutinize all your changes. Really appreciate your effort so far!

kbilsted commented on StatePrinter.Tests/IntegrationTests/DictionaryTest.cs:
```diff
@@ -803,13 +836,13 @@ public void Dictionary_complexKeySimpleValue()
     <Element type='KeyValuePair(Person, Int32)'>
         <key type='Person'>
-            <Name>""Douglas""</Name>
+            <Name>Douglas</Name>
```
I recall wanting the " in there to specify that we are dealing with a string - but I totally agree with you it has no place in neither json nor xml - perhaps not even in the "CurlyBraceStyle" - it was supposed to be a "best readable" format for programmers.. I think your change is an improvement

romkatv replied:
I think quotes in the CurlyBraceStyle are useful, especially in single-line mode. Compare these:
  { Address = "Zurich, Switzerland" Name = "Roman" }
vs
  { Address = Zurich, Switzerland Name = Roman }

In XML style quotes aren't necessary because each value is already delimited by XML tags.

Right now the CurlyBraceStyle looks kinda like C# while JSON and XML looks kinda like JSON and XML respectively. That's nice.

By the way, there is one thing I forgot to mention, which is related. JSON formatter used to produce strings like this:

  {42: 1337}

This isn't legal JSON -- object fields can't be numbers. I changed it to this:

  {"42": "1337"}

kbilsted commented on StatePrinter/Introspection/IntroSpector.cs:
```diff
@@ -201,15 +203,14 @@ private bool IntrospectIEnumerable(object source, Field field)
+            tokens.Add(StartList);
 
-            int i = 0;
+            long i = 0;
```
how big a graph are you planning on dumping?;) Our memory consumption will increase with this change. How about using a uint? unsigned i guess is nicer in that the index can not be negative - just like an unsigned cant be negative

romkatv replied:
Sorry, force of habit. In my day job the assumption of fewer than 2B elements per sequence doesn't hold.

I reverted it back to int. There are probably a few more places where StatePrinter won't handle huge sequences so there is no point in using long just in this place.

kbilsted commented on StatePrinter/Introspection/Token.cs:
```diff
@@ -89,9 +89,14 @@ public override int GetHashCode()
-            var name =  Field == null ? "" : Field.Name;
-            var simpleKeyInArrayOrDictionary = Field == null ? "" : Field.SimpleKeyInArrayOrDictionary;
-            return string.Format("{0}: {1} {2} = {3}({4})", Tokenkind, FieldType, name, Value, simpleKeyInArrayOrDictionary);
+            string name = Field == null ? null : Field.Name;
```
perhaps an early return if field = null makes this code less convoluted

romkatv replied:
Good idea. Done.

kbilsted commented on StatePrinter/Introspection/Token.cs:
```diff
@@ -89,9 +89,14 @@ public override int GetHashCode()
+                key = Field.Key;
+            else if (key == null && Field != null && Field.Index.HasValue)
+                key = Field.Index.Value.ToString();
+            string subscript = key == null ? "" : "[" + key + "]";
```
subscript never used

romkatv replied:
Oops. Fixed.

kbilsted commented on StatePrinter/OutputFormatters/JsonStyle.cs:
```diff
@@ -118,104 +121,62 @@ string MakeString(List<Token> tokens, Dictionary<Reference, string> referencePat
-                    sb.AppendFormatLine("{{");
-                    sb.Indent();
+                case TokenType.StartDict:
+                    if (next == TokenType.EndScope || next == TokenType.EndDict)
```
in case of `(next == TokenType.EndScope || next == TokenType.EndDict)`  method `OptionalComma()` always return `""` - what are you trying to achieve with the `if`?

romkatv replied:
We call `OptionalComma(tokens, pos + 1)` here (note the `+1`). This `if` is responsible for single-line printing of empty objects.

kbilsted commented on StatePrinter.Tests/ExamplesForDocumentation/ExampleListAndArrays.cs:
```diff
@@ -41,23 +41,25 @@ public void ExampleListAndArraysAlternative()
-    Allocation = 880
-    Price = 11
-    Share = 50
+    new VendorAllocation()
```
Is there a way we can insert the index in such a way as it reads more natural than in the old format? I have a plan on making stateprinter a tool for helping with debugging states at certain points in the code - and thus seeing the possition may be nice. Thats why they were included.

romkatv replied:
I thought it would be nice to produce more C#-like code, that's why I removed the indices.

Since you prefer to have them, I put them back. I don't have strong feelings either way and I see how they can be useful.

kbilsted replied:
Well nothing prevents us from creating yet another output format that is less verbose ;)

kbilsted commented on StatePrinter.Tests/IntegrationTests/DictionaryTest.cs:
```diff
@@ -76,21 +76,41 @@ public void Setup()
+    new Token()
+    {
+        Tokenkind = FieldnameWithTypeAndReference
+        FieldType = new RuntimeType()
```
what is this RuntimeType ?

romkatv replied:
That's what `object.GetType()` returns. You can see several `RuntimeType()` instances in the current code of this file (prior to my changes).

kbilsted commented on StatePrinter.Tests/IntegrationTests/DictionaryTest.cs:
```diff
@@ -168,219 +211,263 @@ public void Dictionary_int_int()
+        Value = null
+        ReferenceNo = null
+    }
+    -> 4
```
-> 5 -> 6 -> 5 becomes -> 9 -> 4 -> 4   is that correct?

romkatv replied:
This is correct. The list contains different elements now and the reference numbering is different.

kbilsted commented on StatePrinter.Tests/IntegrationTests/DictionaryTest.cs:
```diff
@@ -404,9 +491,12 @@ public void Setup()
+                var expected = @"new Dictionary<Int32, Int32>()
+{
+    {1, 2}
+    {2, 4}
```
:+1:

kbilsted commented on StatePrinter.Tests/IntegrationTests/DictionaryTest.cs:
```diff
@@ -425,34 +518,42 @@ public void Dictionary_int_int_withRootName()
-[0] = new KeyValuePair<Person, Int32>()
 {
-    key = new Person()
+    new KeyValuePair<Person, Int32>()
```
as mentioned before, we are now loosing the ability to see the index position

romkatv replied:
How about now?

kbilsted commented on StatePrinter.Tests/IntegrationTests/DictionaryTest.cs:
```diff
@@ -524,11 +633,11 @@ public void Setup()
-    { 3 : 6 }
-]";
+                var expected = @"{
+    ""1"": 2, 
```
ah this is the json keys that cannot be numbers..

romkatv replied:
Yeap.

kbilsted commented on StatePrinter.Tests/IntegrationTests/DictionaryTest.cs:
```diff
@@ -586,19 +693,18 @@ public void Dictionary_complexKeySimpleValue()
-        ""_key"" : 3,
-        ""_value"" : 6
-    }
+        ""_key"": 3, 
```
why is there a space at the end of the line? looks wrong

romkatv replied:
Nice catch! Fixed.

kbilsted commented on StatePrinter.Tests/IntegrationTests/DictionaryTest.cs:
```diff
@@ -698,44 +800,40 @@ public void Dictionary_empty_int_int()
-    <Enumeration>
-    <ROOT type='DictionaryEntry'>
+                var expected = @"<Root type='Hashtable'>
+    <Element type='DictionaryEntry'>
```
:+1:

kbilsted commented on StatePrinter.Tests/IntegrationTests/ObjectGraphsTest.cs:
```diff
@@ -173,37 +164,47 @@ public void MegaCyclicGraph_Curlybrace()
                 Children = new List<Human>()
-                Father =  -> 2
+                {
+                    -> 0
```
this is really hard to read - if it wasn't for the old output in the diff, I would not know it was the children.

romkatv replied:
How about now?

kbilsted commented:
hi @romkatv  I hope I haven't discouraged you with my many many comments. I think the problem is that the PR is spanning too many things perhaps. I think it would be easier to understand and discuss if it was broken down into more PR's and with a branch for each PR. But im not sure how familiar you are with git and branching? if you are familiar with branching I'd prefer we create a ticket for one of the things in here, and then create a branch for that issue and then a PR for that branch. Then creating release notes is really easy. and we can easier discuss and close each PR. I would leave this master unchanged for now so that we don't loose all the comments. how does all this sound? too much work on your part? if so let me know..  its not like it want to prevent your changes going in ;) its rather the opposite ;)

thanks again for the huge amount of work so far

romkatv commented:
I addressed all your comments. Please take a another look.

My apologies for creating such a large PR. Seems like we are converging in our discussion. Perhaps you could accept it as a whole? I can't find the motivation for splitting it up.

kbilsted commented:
no problem not splitting it up. I need to review the changes compared to my master.. I haven't figured out a nice way to doing this. I guess I can `git reset --soft master` and then check the changes. But surely there is a better way ;) will look into it sooon

kbilsted commented on StatePrinter.Tests/IntegrationTests/DictionaryTest.cs:
```diff
@@ -524,23 +647,33 @@ public void Setup()
-    { 2 : 4 },
-    { 3 : 6 }
-]";
+                var expected = @"{
```
why are the keys inside `"` now that it is an int -> int dictionary?

romkatv commented:
Thanks for reviewing and accepting the PR. It was a tough one.

When do you expect to push a new version to NuGet?

kbilsted commented:
I want to rename the main class for VBA users. Also thinking about lookin into some performance tests. Your pr made one test go from 25 to 23 seconds..that's 10% :)

kbilsted commented:
The configuration can do with caching the unmatched fields.  Maybe you want to do that?

romkatv commented:
Not right now. I'm busy with a few other projects.

kbilsted commented:
Also I will create 4 issues found on the basis of your pr  that we may want to solve first.will create those issues today

kbilsted commented:
No problem. Many thanks for the pr. Will make a new release within a week
